pkg/aws/g8scontrolplane/mutate_g8scontrolplane.go

```diff
@@ -66,89 +79,187 @@ func (m *Mutator) Mutate(request *admissionv1.AdmissionRequest) ([]mutator.Patch
-		replicas = aws.DefaultMasterReplicas
-		update := func() error {
-			ctx := context.Background()
 
-			// We fetch the AWSControlPlane CR.
-			awsControlPlane := &infrastructurev1alpha2.AWSControlPlane{}
-			{
-				m.Log("level", "debug", "message", fmt.Sprintf("Fetching AWSControlPlane %s", g8sControlPlaneNewCR.Name))
-				err := m.k8sClient.CtrlClient().Get(ctx,
-					types.NamespacedName{Name: g8sControlPlaneNewCR.GetName(),
-						Namespace: namespace},
-					awsControlPlane)
-				if err != nil {
-					return microerror.Maskf(notFoundError, "failed to fetch AWSControlplane: %v", err)
-				}
-			}
-			// If there is an AWSControlPlane, the default replicas match the number of AZs
-			replicas = len(awsControlPlane.Spec.AvailabilityZones)
-			// If there is an AWSControlplane, we get its infrastructure reference
-			infrastructureCRRef, err = reference.GetReference(infrastructurev1alpha2scheme.Scheme, awsControlPlane)
-			if err != nil {
-				return microerror.Mask(err)
-			}
+	patch, err = m.MutateReleaseVersion(*g8sControlPlaneCR)
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+	releaseVersion, err := releaseVersion(g8sControlPlaneCR, patch)
+	if err != nil {
+		return nil, microerror.Maskf(parsingFailedError, "unable to parse release version from G8sControlPlane")
+	}
+	result = append(result, patch...)
 
-			// If the availability zones need to be updated from 1 to 3, we do it here
-			{
-				if aws.IsHAVersion(releaseVersion) && isUpdateFromSingleToHA(g8sControlPlaneNewCR, g8sControlPlaneOldCR) && len(awsControlPlane.Spec.AvailabilityZones) == 1 {
-					m.Log("level", "debug", "message", fmt.Sprintf("Updating AWSControlPlane AZs for HA %s", awsControlPlane.Name))
-					awsControlPlane.Spec.AvailabilityZones = m.getHAavailabilityZones(awsControlPlane.Spec.AvailabilityZones[0], m.validAvailabilityZones)
-					err := m.k8sClient.CtrlClient().Update(ctx, awsControlPlane)
-					if err != nil {
-						return microerror.Mask(err)
-					}
-				}
-				return nil
-			}
-		}
-		b := backoff.NewMaxRetries(3, 1*time.Second)
-		err := backoff.Retry(update, b)
+	// We try to fetch the AWSControlPlane belonging to the G8sControlPlane here.
+	availabilityZones := 0
+	awsControlPlane, err := m.fetchAWSControlPlane(*g8sControlPlaneCR)
+	if IsNotFound(err) {
 		// Note that while we do log the error, we don't fail if the AWSControlPlane doesn't exist yet. That is okay because the order of CR creation can vary.
-		if IsNotFound(err) {
-			m.Log("level", "debug", "message", fmt.Sprintf("No AWSControlPlane %s could be found: %v", g8sControlPlaneNewCR.Name, err))
-		} else if err != nil {
-			return nil, err
+		m.Log("level", "debug", "message", fmt.Sprintf("No AWSControlPlane %s could be found: %v", g8sControlPlaneCR.GetName(), err))
+	} else if err != nil {
+		return nil, microerror.Mask(err)
+	} else {
+		// This defaulting is only done when the awscontrolplane exists
+		availabilityZones = len(awsControlPlane.Spec.AvailabilityZones)
```

Comment:
We have to remember the number of AZs because it's needed for the replica defaulting later.